presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,534 @@
+import com.facebook.presto.spi.storage.TempStorageHandle;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.TreeRangeMap;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.units.DataSize;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.facebook.presto.execution.buffer.BufferResult.emptyResults;
+import static com.facebook.presto.execution.buffer.BufferState.FINISHED;
+import static com.facebook.presto.execution.buffer.BufferState.NO_MORE_PAGES;
+import static com.facebook.presto.execution.buffer.BufferState.OPEN;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+@SuppressWarnings("UnstableApiUsage")
+public class SpoolingOutputBuffer
+        implements OutputBuffer
+{
+
+    private final OutputBuffers outputBuffers;
+    private final StateMachine<BufferState> state;
+    private final String taskInstanceId;
+    private final TempDataOperationContext context;
+    private final TempStorage tempStorage;
+    private final DataSize threshold;
+
+    private final AtomicBoolean noMorePages = new AtomicBoolean();
+    private final AtomicInteger startPage = new AtomicInteger();
+    private final AtomicLong memorySequenceId = new AtomicLong();
+    private final AtomicLong sequenceId = new AtomicLong();
+
+    private final AtomicLong totalBufferedBytes = new AtomicLong();
+    private final AtomicLong totalBufferedPages = new AtomicLong();
+    private final AtomicLong totalPagesAdded = new AtomicLong();
+    private final AtomicLong totalRowsAdded = new AtomicLong();
+    private final AtomicLong totalPagesRemaining = new AtomicLong();
+
+    private final Closer closer = Closer.create();
+
+    private final LinkedList<SerializedPage> pages = new LinkedList<>();
+    private final OutputBuffers.OutputBufferId outputBufferId = new OutputBuffers.OutputBufferId(0);
+    private final RangeMap<Long, TempStorageHandle> handleRangeMap = TreeRangeMap.create();
```

Comment:
* Let's explicitly model that we are serializing batches of pages using an inner class.
* This class can have the range of sequence ids that it represents
* Instead of keying off of this map using the sequence ID, I would simply use a `Queue`.  When we consume, we can keep polling the queue until the sequence ids are in the desired range.